src/boot/efi/bcd.c

```diff
@@ -0,0 +1,286 @@
+                return NULL;
+
+        SubkeyFast *subkey = (SubkeyFast *) (bcd + offset);
+        if (subkey->sig != SIG_SUBKEY_FAST)
+                return NULL;
+
+        if ((UINT64) offset + offsetof(SubkeyFast, entries) + sizeof(struct SubkeyFastEntry[subkey->n_entries]) > bcd_len)
+                return NULL;
+
+        for (UINT16 i = 0; i < subkey->n_entries; i++) {
+                if (strncasecmpa(name, subkey->entries[i].name_hint, sizeof(subkey->entries[i].name_hint)) != 0)
+                        continue;
+
+                Key *key = get_key(bcd, bcd_len, subkey->entries[i].key_offset, name);
+                if (key)
+                        return key;
+        }
+
+        return NULL;
+}
+
+/* We use NUL as registry path separators for convenience. To start from the root, begin
+ * name with a NUL. Name must end with two NUL. */
+static Key *get_key(const UINT8 *bcd, UINT32 bcd_len, UINT32 offset, const CHAR8 *name) {
+        assert(bcd);
+        assert(name);
+
+        if ((UINT64) offset + sizeof(Key) > bcd_len)
+                return NULL;
+
+        Key *key = (Key *) (bcd + offset);
+        if (key->sig != SIG_KEY)
+                return NULL;
+
+        if ((UINT64) offset + offsetof(Key, key_name) + sizeof(CHAR8[key->key_name_len]) > bcd_len)
+                return NULL;
+
+        if (*name) {
+                if (strncasecmpa(name, key->key_name, key->key_name_len) == 0 && !name[key->key_name_len])
+                        name += key->key_name_len;
+                else
+                        return NULL;
+        }
+
+        name++;
+        return *name ? get_subkey(bcd, bcd_len, key->subkeys_offset, name) : key;
+}
+
+static KeyValue *get_key_value(const UINT8 *bcd, UINT32 bcd_len, const Key *key, const CHAR8 *name) {
+        assert(bcd);
+        assert(key);
+        assert(name);
+
+        if (key->n_key_values == 0)
+                return NULL;
+
+        if ((UINT64) key->key_values_offset + sizeof(UINT32[key->n_key_values]) > bcd_len)
+                return NULL;
+
+        UINT32 *key_value_list = (UINT32 *) (bcd + key->key_values_offset);
+        for (UINT32 i = 0; i < key->n_key_values; i++) {
+                UINT32 offset = *(key_value_list + i);
+
+                if ((UINT64) offset + sizeof(KeyValue) > bcd_len)
+                        continue;
+
+                KeyValue *kv = (KeyValue *) (bcd + offset);
+                if (kv->sig != SIG_KEY_VALUE)
+                        continue;
+
+                if ((UINT64) offset + offsetof(KeyValue, name) + kv->name_len> bcd_len)
+                        continue;
+
+                /* If most significant bit is set, data is stored in data_offset itself, but
+                 * we are only interested in UTF16 strings. The only strings that could fit
+                 * would have just one char in it, so let's not bother with this. */
+                if (FLAGS_SET(kv->data_size, UINT32_C(1) << UINT32_C(31)))
```

Comment:
That second UINT32_C() is unnecessary and specifying the width is pretty much pointless: it has no effect on the output.